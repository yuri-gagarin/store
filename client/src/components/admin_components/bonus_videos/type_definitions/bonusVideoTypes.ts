// api actions types //
export interface IBonusVideoServerResData {
  responseMsg: string;
  bonusVideo?: IBonusVideoData;
  newBonusVideo?: IBonusVideoData;
  editedBonusVideo?: IBonusVideoData;
  deletedBonusVideo?: IBonusVideoData;
  bonusVideos?: IBonusVideoData[];
}
export interface IBonusVideoServerRes {
  data: IBonusVideoServerResData
}
export type ClientBonusVideoData = {
  description?: string;
  youTubeURL?: string;
  vimeoURL?: string;
}
// form types //
export type BonusVideoData = {
  description?: string;
  youTubeURL?: string;
  vimeoURL?: string;
}

export type FormState = {
  description?: string;
  youTubeURL?: string;
  vimeoURL?: string;
}