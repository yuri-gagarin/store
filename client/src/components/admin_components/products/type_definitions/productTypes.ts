// APIProduct actions types //
export interface IProductImgServerResData {
  responseMsg: string;
  newProductImage?: IProductImgData;
  deletedProductImage?: IProductImgData;
  updatedProduct: IProductData;
}
export interface IProductServerRes {
  data: IProductServerResData
}
export interface IProductImgServerRes {
  data: IProductImgServerResData;
}
export interface IProductServerResData {
  responseMsg: string;
  product?: IProductData;
  newProduct?: IProductData;
  editedProduct?: IProductData;
  deletedProduct?: IProductData;
  products?: IProductData[];
}
export type ClientProductData = {
  name: string;
  description: string;
  price: string;
  images: IProductImgData[]
}
// ProductFormHolder and ProductForm types //
export type ProductData = {
  name: string;
  description: string;
  details: string;
  price: string;
  images: IProductImgData[];
}
export type ProductFormState = {
  name: string;
  description: string;
  details: string;
  price: string;
}