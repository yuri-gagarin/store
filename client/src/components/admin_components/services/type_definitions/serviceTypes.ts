// APIServiceActions types //
export interface IServiceImgServerResData {
  responseMsg: string;
  newServiceImage?: IServiceImgData;
  deletedServiceImage?: IServiceImgData;
  updatedService: IServiceData;
}
export interface IServiceImgServerRes {
  data: IServiceImgServerResData;
}
export interface IServiceServerResData {
  responseMsg: string;
  service?: IServiceData;
  newService?: IServiceData;
  editedService?: IServiceData;
  deletedService?: IServiceData;
  services?: IServiceData[];
}
export interface IServiceServerRes {
  data: IServiceServerResData
}
export type ClientServiceData = {
  name: string;
  description: string;
  price: string;
  images: IServiceImgData[]
}
// 