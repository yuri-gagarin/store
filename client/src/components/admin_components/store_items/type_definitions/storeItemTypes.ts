// StoreItem - APIStoreItemActions types //
export interface IStoreItemImgServerResData {
  responseMsg: string;
  newStoreItemImage?: IStoreItemImgData;
  deletedStoreItemImage?: IStoreItemImgData;
  updatedStoreItem: IStoreItemData;
}

export interface IStoreItemImgServerRes {
  data: IStoreItemImgServerResData;
}

export interface IStoreItemServerResData {
  responseMsg: string;
  numberOfItems: string;
  storeItem?: IStoreItemData;
  newStoreItem?: IStoreItemData;
  editedStoreItem?: IStoreItemData;
  deletedStoreItem?: IStoreItemData;
  storeItems?: IStoreItemData[];
}

export interface IStoreItemServerRes {
  data: IStoreItemServerResData
}

export type ClientStoreItemData = {
  storeId: string;
  storeName: string;
  name: string;
  description: string;
  details: string;
  price: string;
  categories: string[];
  images: IStoreItemImgData[]
}

export type StoreItemQueryPar = {
  storeName?: string;
  storeId?: string;
  limit?: string;
  date?: string;
  price?: string;
  name?: string;
}

// StoreItem form and holder types //
export type StoreItemFormState = {
  storeId: string;
  storeName: string;
  name: string;
  description: string;
  details: string;
  price: string;
  categories: string[];
}
export type StoreDropdownData = {
  storeId: string;
  storeName: string;
}
export type StoreItemData = {
  storeId: string,
  storeName: string,
  name: string;
  description: string;
  details: string;
  price: string;
  images: IStoreItemImgData[];
  categories: string[];
}
// StoreItemFormDropdown types //
export type StoreItemFormStoreDropdownData = {
  key: string;
  text: string;
  value: string;
}
export type StoreItemFormStoreDropdownState = {
  dropdownData: StoreItemFormStoreDropdownData[];
  disabled: boolean;
}
// StoreItemsControlls component types //
export type DropdownData = {
  key: string;
  text: string;
  value: string;
}
export type DropdownState = {
  loading: boolean;
  disabled: boolean;
  data: DropdownData[];
}
