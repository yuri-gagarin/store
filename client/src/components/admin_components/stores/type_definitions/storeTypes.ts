export interface IStoreImgServerRes {
  data: IStoreImgServerResData;
}
export interface IStoreImgServerResData {
  responseMsg: string;
  newStoreImage?: IStoreImgData;
  deletedStoreImage?: IStoreImgData;
  updatedStore: IStoreData;
}
export interface IStoreServerResponse {
  data: IStoreServerResData
}
export interface IStoreServerResData {
  responseMsg: string;
  store?: IStoreData;
  newStore?: IStoreData;
  editedStore?: IStoreData;
  deletedStore?: IStoreData;
  stores?: IStoreData[];
}

export type ClientStoreData = {
  title: string;
  description: string;
  images: IStoreImgData[]
}
export type StoreQuery = {
  storeName?: string;
  limit?: string;
}

// StoreFormHolder and StoreForm components //
export type StoreFormState = {
  title: string;
  description: string;
}