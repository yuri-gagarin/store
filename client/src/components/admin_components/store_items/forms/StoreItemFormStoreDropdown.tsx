import React, {  useEffect, useState, useContext } from "react";
import { Dropdown, DropdownProps } from "semantic-ui-react";
// helpers //
import { capitalizeString } from "../../../helpers/displayHelpers";
// types //
import { StoreDropdownData, StoreItemFormStoreDropdownData, StoreItemFormStoreDropdownState } from "../type_definitions/storeItemTypes";

interface Props {
  activeStores: StoreDropdownData[];
  setStoreItemStore: (data: StoreDropdownData) => void;
}

const StoreItemFormStoreDropdown: React.FC<Props> = ({ activeStores, setStoreItemStore }): JSX.Element => {

  const [ dropdownState, setDropdownState ] = useState<StoreItemFormStoreDropdownState>({ disabled: true, dropdownData: []});

  const handleSearchChange = (e: React.SyntheticEvent, data: DropdownProps): void => {
    const storeName = data.value as string;
    const selectedStore: StoreItemFormStoreDropdownData = dropdownState.dropdownData.filter((stores) => {
      return stores.value === storeName;
    })[0];
    setStoreItemStore({ storeName: selectedStore.value, storeId: selectedStore.key });
  }
  
  useEffect(() => {
    if (activeStores.length > 0) {
      const dropdownData: StoreItemFormStoreDropdownData[] = activeStores.map((store) => {
        return {
          key: store.storeId,
          text: capitalizeString(store.storeName),
          value: store.storeName
        }
      });
      setDropdownState({
        disabled: false,
        dropdownData: dropdownData
      });
    } else {
      setDropdownState({
        disabled: true,
        dropdownData: []
      });
    }
  }, [ activeStores ]);

  return (
    <Dropdown
      placeholder={"Select a Store"}
      selection
      onChange={handleSearchChange}
      options={dropdownState.dropdownData}
      disabled={dropdownState.disabled}
    />
  )
};

export default StoreItemFormStoreDropdown;