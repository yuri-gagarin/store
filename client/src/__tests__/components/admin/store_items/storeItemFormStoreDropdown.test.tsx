import React from "react";
import { mount, ReactWrapper } from "enzyme";
// component imports //
import StoreItemFormStoreDropdown from "../../../../components/admin_components/store_items/forms/StoreItemFormStoreDropdown";
import { Dropdown } from "semantic-ui-react";
import { StoreDropdownData } from "../../../../components/admin_components/store_items/type_definitions/storeItemTypes";

describe("'StoreItemFormStoreDropdown' component render tests", () => {

  describe("'StoreItemFormStoreDropdown' component in default local state", () => {
    let wrapper: ReactWrapper;

    beforeAll(() => {
      wrapper = mount(
        <StoreItemFormStoreDropdown activeStores={[]} setStoreItemStore={jest.fn}/>
      )
    });

    it("Should properly render without errors", () => {
      expect(wrapper.find(Dropdown)).toMatchSnapshot();
    });
    it("'Dropdown' component should be disabled by default", () => {
      const dropdown = wrapper.find(Dropdown);
      expect(dropdown.props().disabled).toEqual(true);
    });
  });
  
  describe("'StoreItemFormStoreDropdown' component render tests with {activeStores} 'prop' passed in", () => {
    let wrapper: ReactWrapper; 
    let mockStores: StoreDropdownData[]

    beforeAll(() => {
      mockStores = [
        { storeId: "1", storeName: "first" },
        { storeId: "2", storeName: "second" },
        { storeId: "3", storeName: "third" }
      ];
      // mount and pass props //
      wrapper = mount(
        <StoreItemFormStoreDropdown 
          activeStores={mockStores} 
          setStoreItemStore={jest.fn}
        />
      );
    });

    it("Should properly render without errors", () => {
      expect(wrapper.find(Dropdown)).toMatchSnapshot();
    });
    it("'Dropdown' component should be NOT be disabled", () => {
      const dropdown = wrapper.find(Dropdown);
      expect(dropdown.props().disabled).toEqual(false);
    });
    it("Should render a correct number of 'Dropdown' child components", () => {
      const dropdownItems = wrapper.find(Dropdown.Item);
      expect(dropdownItems.length).toEqual(mockStores.length);
    });
  });
  
});