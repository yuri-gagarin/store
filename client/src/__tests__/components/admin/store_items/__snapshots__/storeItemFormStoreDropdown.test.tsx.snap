// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`'StoreItemFormStoreDropdown' component render tests 'StoreItemFormStoreDropdown' component in default local state Should properly render without errors 1`] = `
<Dropdown
  additionLabel="Add "
  additionPosition="top"
  closeOnBlur={true}
  closeOnEscape={true}
  deburr={false}
  disabled={true}
  icon="dropdown"
  minCharacters={1}
  noResultsMessage="No results found."
  onChange={[Function]}
  openOnFocus={true}
  options={Array []}
  placeholder="Select a Store"
  renderLabel={[Function]}
  searchInput="text"
  selectOnBlur={true}
  selectOnNavigation={true}
  selection={true}
  wrapSelection={true}
>
  <Ref
    innerRef={
      Object {
        "current": <div
          aria-disabled="true"
          aria-expanded="false"
          class="ui disabled selection dropdown"
          role="listbox"
          tabindex="-1"
        >
          <div
            aria-atomic="true"
            aria-live="polite"
            class="default text"
            role="alert"
          >
            Select a Store
          </div>
          <i
            aria-hidden="true"
            class="dropdown icon"
          />
          <div
            class="menu transition"
          />
        </div>,
      }
    }
  >
    <RefFindNode
      innerRef={
        Object {
          "current": <div
            aria-disabled="true"
            aria-expanded="false"
            class="ui disabled selection dropdown"
            role="listbox"
            tabindex="-1"
          >
            <div
              aria-atomic="true"
              aria-live="polite"
              class="default text"
              role="alert"
            >
              Select a Store
            </div>
            <i
              aria-hidden="true"
              class="dropdown icon"
            />
            <div
              class="menu transition"
            />
          </div>,
        }
      }
    >
      <div
        aria-disabled={true}
        aria-expanded={false}
        className="ui disabled selection dropdown"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        role="listbox"
        tabIndex={-1}
      >
        <div
          aria-atomic={true}
          aria-live="polite"
          className="default text"
          role="alert"
        >
          Select a Store
        </div>
        <Icon
          as="i"
          className=""
          name="dropdown"
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="dropdown icon"
            onClick={[Function]}
          />
        </Icon>
        <DropdownMenu>
          <div
            className="menu transition"
          />
        </DropdownMenu>
      </div>
    </RefFindNode>
  </Ref>
</Dropdown>
`;

exports[`'StoreItemFormStoreDropdown' component render tests 'StoreItemFormStoreDropdown' component render tests with {activeStores} 'prop' passed in Should properly render without errors 1`] = `
<Dropdown
  additionLabel="Add "
  additionPosition="top"
  closeOnBlur={true}
  closeOnEscape={true}
  deburr={false}
  disabled={false}
  icon="dropdown"
  minCharacters={1}
  noResultsMessage="No results found."
  onChange={[Function]}
  openOnFocus={true}
  options={
    Array [
      Object {
        "key": "1",
        "text": "First",
        "value": "first",
      },
      Object {
        "key": "2",
        "text": "Second",
        "value": "second",
      },
      Object {
        "key": "3",
        "text": "Third",
        "value": "third",
      },
    ]
  }
  placeholder="Select a Store"
  renderLabel={[Function]}
  searchInput="text"
  selectOnBlur={true}
  selectOnNavigation={true}
  selection={true}
  wrapSelection={true}
>
  <Ref
    innerRef={
      Object {
        "current": <div
          aria-disabled="false"
          aria-expanded="false"
          class="ui selection dropdown"
          role="listbox"
          tabindex="0"
        >
          <div
            aria-atomic="true"
            aria-live="polite"
            class="default text"
            role="alert"
          >
            Select a Store
          </div>
          <i
            aria-hidden="true"
            class="dropdown icon"
          />
          <div
            class="menu transition"
          >
            <div
              aria-checked="false"
              aria-selected="true"
              class="selected item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                First
              </span>
            </div>
            <div
              aria-checked="false"
              aria-selected="false"
              class="item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                Second
              </span>
            </div>
            <div
              aria-checked="false"
              aria-selected="false"
              class="item"
              role="option"
              style="pointer-events: all;"
            >
              <span
                class="text"
              >
                Third
              </span>
            </div>
          </div>
        </div>,
      }
    }
  >
    <RefFindNode
      innerRef={
        Object {
          "current": <div
            aria-disabled="false"
            aria-expanded="false"
            class="ui selection dropdown"
            role="listbox"
            tabindex="0"
          >
            <div
              aria-atomic="true"
              aria-live="polite"
              class="default text"
              role="alert"
            >
              Select a Store
            </div>
            <i
              aria-hidden="true"
              class="dropdown icon"
            />
            <div
              class="menu transition"
            >
              <div
                aria-checked="false"
                aria-selected="true"
                class="selected item"
                role="option"
                style="pointer-events: all;"
              >
                <span
                  class="text"
                >
                  First
                </span>
              </div>
              <div
                aria-checked="false"
                aria-selected="false"
                class="item"
                role="option"
                style="pointer-events: all;"
              >
                <span
                  class="text"
                >
                  Second
                </span>
              </div>
              <div
                aria-checked="false"
                aria-selected="false"
                class="item"
                role="option"
                style="pointer-events: all;"
              >
                <span
                  class="text"
                >
                  Third
                </span>
              </div>
            </div>
          </div>,
        }
      }
    >
      <div
        aria-disabled={false}
        aria-expanded={false}
        className="ui selection dropdown"
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        role="listbox"
        tabIndex={0}
      >
        <div
          aria-atomic={true}
          aria-live="polite"
          className="default text"
          role="alert"
        >
          Select a Store
        </div>
        <Icon
          as="i"
          className=""
          name="dropdown"
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="dropdown icon"
            onClick={[Function]}
          />
        </Icon>
        <DropdownMenu>
          <div
            className="menu transition"
          >
            <DropdownItem
              active={false}
              key="1"
              onClick={[Function]}
              selected={true}
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
              text="First"
              value="first"
            >
              <div
                aria-checked={false}
                aria-selected={true}
                className="selected item"
                onClick={[Function]}
                role="option"
                style={
                  Object {
                    "pointerEvents": "all",
                  }
                }
              >
                <span
                  className="text"
                >
                  First
                </span>
              </div>
            </DropdownItem>
            <DropdownItem
              active={false}
              key="2"
              onClick={[Function]}
              selected={false}
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
              text="Second"
              value="second"
            >
              <div
                aria-checked={false}
                aria-selected={false}
                className="item"
                onClick={[Function]}
                role="option"
                style={
                  Object {
                    "pointerEvents": "all",
                  }
                }
              >
                <span
                  className="text"
                >
                  Second
                </span>
              </div>
            </DropdownItem>
            <DropdownItem
              active={false}
              key="3"
              onClick={[Function]}
              selected={false}
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
              text="Third"
              value="third"
            >
              <div
                aria-checked={false}
                aria-selected={false}
                className="item"
                onClick={[Function]}
                role="option"
                style={
                  Object {
                    "pointerEvents": "all",
                  }
                }
              >
                <span
                  className="text"
                >
                  Third
                </span>
              </div>
            </DropdownItem>
          </div>
        </DropdownMenu>
      </div>
    </RefFindNode>
  </Ref>
</Dropdown>
`;
